### PR TITLE
Add NixOS

### PR DESCRIPTION
**Pull request by gianarb**

### CODE REVIEW

File: http/nixos-20-09/install.sh [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+a
+w
+
+FDISK
+
+# Create filesystem
+mkfs.ext4 -j -L nixos /dev/sda1
+
+partprobe
```
Comment (mmlb) [inline http/nixos-20-09/install.sh:22]: partprobe should come before mkfs.

**Review by mmlb — CHANGES_REQUESTED** [review]

File: http/nixos-20-09/postinstall.sh [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+for x in $(seq 0 2) ; do
+  nix-env --delete-generations old
+  nix-collect-garbage -d
+done
+
+# Remove install ssh key
+rm -rf /root/.ssh /root/.packer_http
+
+if [[ "${PACKER_BUILDER_TYPE}" == "qemu" ]] ; then
+  echo "skipping disk zero out!"
```
Comment (mmlb) [inline http/nixos-20-09/postinstall.sh:17]: lies!

**Review by mmlb — CHANGES_REQUESTED** [review]

File: http/nixos-20-09/postinstall.sh [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+else
+  echo "zeroing out the disk..."
+
+  # Zero out the disk (for better compression)
+  dd if=/dev/zero of=/EMPTY bs=1M
```
Comment (mmlb) [inline http/nixos-20-09/postinstall.sh:22]: `cat /dev/zero >/EMPTY` is usually faster than all the extra overhead of `dd`

File: http/nixos-20-09/install.sh (hunk 1 above)
Comment (gianarb) [reply http/nixos-20-09/install.sh:22]: it is not needed anymore :D it was a mistake on my end!

**Comment by mmlb** [comment]
lgtm now, and oops I didn't notice this was a Draft PR.

**Comment by gianarb** [comment]
yes, it is a draft because I didn't even run it with Tinkerbell yet xD.

And I think it will be cool to add something sweet thing like ` import ./custom.nix` that can be overwritten using the [cexec or writefile action](https://artifacthub.io/packages/tbaction/tinkerbell-community/writefile) in order to do something similar to what EM does passing a NixOS configuration via "user data"

**Comment by mmlb** [comment]
Yes that would be nice

**Comment by thebsdbox** [comment]
image

I get above and then a kernel panic :(

**Comment by gianarb** [comment]
Probably there is some kernel module missed for your NUC. I can't tell on
virtual box I was able to get it working

**Comment by thebsdbox** [comment]
Good point.

**Comment by gianarb** [comment]
I am gonna close this for now. We generate the hardware-configuration.nix file from QEMU, but the workflow has to regenerate the NixOS configuration and run nixos-install, probably with `cexec` action or similar otherwise the kernel is configured for qemu and not for the end machine.

https://nixos.org/manual/nixos/stable/index.html#sec-installation

I will leave it for somebody else more motivated than me

**Pull request closed without merging**